cxx/distributions/normal.hh

```diff
@@ -37,19 +59,45 @@ public:
         var -= (x - mean) * (x - old_mean);
     }
 
+    double sprime() const {
+      // r' = r + N
+      // m' = (r m + N mean) / (r + N)
+      // C = N (var + mean^2)
+      // s' = s + C + r m^2 - r' m' m'
+      double mdelta = r * (m - mean) / (r + N);
+      double mprime = mean + mdelta;
+      return s + r * (m * m - mprime * mprime)
+          + N * (var - 2 * mean * mdelta - mdelta * mdelta);
+    }
+
     double logp(double x) const {
-        double y = (x - mean);
-        return -0.5 * (y * y / var + log(var) + log(M_2PI));
+      return -0.5 * N * log(M_2PI)
+          + logZ(r + N, v + N, sprime())
+          - logZ(r, v, s);
     }
 
     double logp_score() const {
-        // TODO(thomaswc): This.
-        return 0.0;
+      double sp = sprime();
+      double mdelta = r * (m - mean) / (r + N);
+      return -logZ(r + N, v + N, sp)
+          - 0.5 * (v + N - 1) * log(var)
+          - 0.5 * ((r + N) * mdelta * mdelta + sp)/ var;
     }
```

Comment:
The reference implementation of this method from Feras looks close to your `logp`: https://github.com/probcomp/cgpm/blob/56a481829448bddc9cdfebd42f65023287d5b7c7/src/primitives/normal.py#L176-L181

Possibly what you have here is also right, but I'd like to understand the equivalence. (Happy to put in some more time tomorrow to do that.)

Reply:
Fixed.